File: src/main/java/com/demonwav/mcdev/platform/mcp/at/AtFile.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mcp.at;

import com.demonwav.mcdev.asset.PlatformAssets;
import com.intellij.extapi.psi.PsiFileBase;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.psi.FileViewProvider;
import javax.swing.Icon;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class AtFile extends PsiFileBase {
    public AtFile(@NotNull FileViewProvider viewProvider) {
        super(viewProvider, AtLanguage.getInstance());
    }

    @NotNull
    @Override
    public FileType getFileType() {
        return AtFileType.getInstance();
    }

    @Override
    public String toString() {
        return "Access Transformer File";
    }

    @Nullable
    @Override
    public Icon getIcon(int flags) {
        return PlatformAssets.MCP_ICON;
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/mcp/at/AtFile.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mcp.at

import com.demonwav.mcdev.asset.PlatformAssets
import com.demonwav.mcdev.facet.MinecraftFacet
import com.demonwav.mcdev.platform.mcp.McpModuleType
import com.intellij.extapi.psi.PsiFileBase
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.module.ModuleUtilCore
import com.intellij.psi.FileViewProvider

class AtFile(viewProvider: FileViewProvider) : PsiFileBase(viewProvider, AtLanguage) {

    init {
        setup()
    }

    private fun setup() {
        if (ApplicationManager.getApplication().isUnitTestMode) {
            return
        }

        val vFile = viewProvider.virtualFile

        val module = ModuleUtilCore.findModuleForFile(vFile, project) ?: return
        val mcpModule = MinecraftFacet.getInstance(module, McpModuleType) ?: return
        mcpModule.addAccessTransformerFile(vFile)
    }

    override fun getFileType() = AtFileType
    override fun toString() = "Access Transformer File"
    override fun getIcon(flags: Int) = PlatformAssets.MCP_ICON
}
